doc/manpages/srptool.1
.de1 NOP
.  it 1 an-trap
.  if \\n[.$] \,\\$*\/
..
.ie t \
.ds B-Font [CB]
.ds I-Font [CI]
.ds R-Font [CR]
.el \
.ds B-Font B
.ds I-Font I
.ds R-Font R
.TH srptool 1 "15 Nov 2023" "3.8.2" "User Commands"
.SH NAME
\f\*[B-Font]srptool\fP
\- GnuTLS SRP tool
.SH SYNOPSIS
\f\*[B-Font]srptool\fP
.\" Mixture of short (flag) options and long options
[\f\*[B-Font]\-flags\f[]]
[\f\*[B-Font]\-flag\f[] [\f\*[I-Font]value\f[]]]
[\f\*[B-Font]\-\-option-name\f[][[=| ]\f\*[I-Font]value\f[]]]
.sp \n(Ppu
.ne 2

All arguments must be options.
.sp \n(Ppu
.ne 2
.SH "DESCRIPTION"
Simple program that emulates the programs in the Stanford SRP (Secure
Remote Password) libraries using GnuTLS.  It is intended for use in  places
where you don't expect SRP authentication to be the used for system users.
.sp
In  brief,  to use SRP you need to create two files. These are the password
file that holds the users and the verifiers associated with  them  and  the
configuration file to hold the group parameters (called tpasswd.conf).
.sp
.SH "OPTIONS"
.TP
.NOP \f\*[B-Font]\-d\f[] \f\*[I-Font]num\f[], \f\*[B-Font]\-\-debug\f[]=\f\*[I-Font]num\f[]
Enable debugging.
This option takes an integer number as its argument.
The value of
\f\*[I-Font]num\f[]
is constrained to being:
.in +4
.nf
.na
in the range 0 through 9999
.fi
.in -4
.sp
Specifies the debug level.
.TP
.NOP \f\*[B-Font]\-i\f[] \f\*[I-Font]num\f[], \f\*[B-Font]\-\-index\f[]=\f\*[I-Font]num\f[]
This option takes an integer number as its argument.
The default
\f\*[I-Font]number\f[]
for this option is:
.ti +4
 3
.sp
.TP
.NOP \f\*[B-Font]\-u\f[] \f\*[I-Font]str\f[], \f\*[B-Font]\-\-username\f[]=\f\*[I-Font]str\f[]
.TP
.NOP \f\*[B-Font]\-p\f[] \f\*[I-Font]str\f[], \f\*[B-Font]\-\-passwd\f[]=\f\*[I-Font]str\f[]
.TP
.NOP \f\*[B-Font]\-s\f[] \f\*[I-Font]num\f[], \f\*[B-Font]\-\-salt\f[]=\f\*[I-Font]num\f[]
This option takes an integer number as its argument.
.TP
.NOP \f\*[B-Font]\-\-verify\f[]
Verifies the password provided against the password file.
.TP
.NOP \f\*[B-Font]\-v\f[] \f\*[I-Font]str\f[], \f\*[B-Font]\-\-passwd\-conf\f[]=\f\*[I-Font]str\f[]
Specify a filename or a PKCS #11 URL to read the CAs from.
.TP
.NOP \f\*[B-Font]\-\-create\-conf\f[]=\f\*[I-Font]str\f[]
Generate a password configuration file.
.sp
This generates a password configuration file (tpasswd.conf)
containing the required for TLS parameters.
.TP
.NOP \f\*[B-Font]\-v\f[] \f\*[I-Font]arg\f[], \f\*[B-Font]\-\-version\f[]=\f\*[I-Font]arg\f[]
Output version of program and exit.  The default mode is `v', a simple
version.  The `c' mode will print copyright information and `n' will
print the full copyright notice.
.TP
.NOP \f\*[B-Font]\-h\f[], \f\*[B-Font]\-\-help\f[]
Display usage information and exit.
.TP
.NOP \f\*[B-Font]\-!\f[], \f\*[B-Font]\-\-more\-help\f[]
Pass the extended usage information through a pager.

.sp
.SH EXAMPLES
To create \fItpasswd.conf\fP which holds the g and n values for SRP protocol
(generator and a large prime), run:
.br
.in +4
.nf
$ srptool \-\-create\-conf /etc/tpasswd.conf
.in -4
.fi
.sp
This command will create \fI/etc/tpasswd\fP and will add user 'test' (you
will also be prompted for a password). Verifiers are stored by default
in the way libsrp expects.
.br
.in +4
.nf
$ srptool \-\-passwd /etc/tpasswd \-\-passwd\-conf /etc/tpasswd.conf \-u test
.in -4
.fi
.sp
.sp
This command will check against a password. If the password matches
the one in \fI/etc/tpasswd\fP you will get an ok.
.br
.in +4
.nf
$ srptool \-\-passwd /etc/tpasswd \-\-passwd\\-conf /etc/tpasswd.conf \-\-verify \-u test
.in -4
.fi
.SH "EXIT STATUS"
One of the following exit values will be returned:
.TP
.NOP 0 " (EXIT_SUCCESS)"
Successful program execution.
.TP
.NOP 1 " (EXIT_FAILURE)"
The operation failed or the command syntax was not valid.
.PP
.SH "SEE ALSO"
gnutls\-cli\-debug (1), gnutls\-serv (1), srptool (1), psktool (1), certtool (1)
.SH "AUTHORS"

.SH "COPYRIGHT"
Copyright (C) 2020-2023 Free Software Foundation, and others all rights reserved.
This program is released under the terms of
the GNU General Public License, version 3 or later
.
.SH "BUGS"
Please send bug reports to: bugs@gnutls.org


ChangeLog
Author: Zoltan Fridrich <zfridric@redhat.com>
Date:   Fri Oct 20 13:43:46 2023 +0200

    Release 3.8.2
    
    Signed-off-by: Zoltan Fridrich <zfridric@redhat.com>

Author: Daiki Ueno <ueno@gnu.org>
Date:   Mon Oct 23 09:26:57 2023 +0900

    auth/rsa_psk: side-step potential side-channel
    
    This removes branching that depends on secret data, porting changes
    for regular RSA key exchange from
    4804febddc2ed958e5ae774de2a8f85edeeff538 and
    80a6ce8ddb02477cd724cd5b2944791aaddb702a.  This also removes the
    allow_wrong_pms as it was used sorely to control debug output
    depending on the branching.
    
    Signed-off-by: Daiki Ueno <ueno@gnu.org>

Author: Ajit Singh <ajeetsinghchahar2@gmail.com>
Date:   Mon Nov 6 15:45:10 2023 +0530

    handshake.c: Fixed a missing goto statement
    
    Signed-off-by: Ajit Singh <ajeetsinghchahar2@gmail.com>

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Sun Nov 5 17:47:44 2023 +0000

    lib: Use correct transport getting error number
    
    For write we need to use transport_send_ptr, not transport_recv_ptr.
    
    Signed-off-by: Frediano Ziglio <freddy77@gmail.com>

Author: Dimitri Papadopoulos <3350651-DimitriPapadopoulos@users.noreply.gitlab.com>
Date:   Thu Oct 26 00:17:33 2023 +0200

    .lgtm.yml: remove LGTM.com configuration file
    
    LGTM.com has been deprecated and replaced by GitHub code analysis:
    https://github.blog/2022-08-15-the-next-step-for-lgtm-com-github-code-scanning/
    
    Signed-off-by: Dimitri Papadopoulos <3350651-DimitriPapadopoulos@users.noreply.gitlab.com>

Author: Daiki Ueno <ueno@gnu.org>
Date:   Wed Oct 18 08:21:05 2023 +0900

    NEWS: mention KTLS support in FreeBSD [ci skip]
    
    Signed-off-by: Daiki Ueno <ueno@gnu.org>

Author: Clemens Lang <cllang@redhat.com>
Date:   Wed Oct 18 10:50:08 2023 +0200

    CONTRIBUTING.md: Fix file name in reviewer checklist
    
    The coding guidelines are in CONTRIBUTING.md, not CONTRIBUTION.md (which
    does not exist).
    
    Signed-off-by: Clemens Lang <cllang@redhat.com>

Author: xuraoqing <xuraoqing@huawei.com>
Date:   Wed Aug 23 11:56:03 2023 +0800

    srptool: rework entry parsing without indexing
    
    This simplifies the loop matching and parsing an SRP entry,
    considering the buffer read with fgets is always NUL-terminated.
    
    Signed-off-by: xuraoqing <xuraoqing@huawei.com>
    Modified-by: Daiki Ueno <ueno@gnu.org>

Author: Daiki Ueno <ueno@gnu.org>
Date:   Wed Oct 18 05:24:33 2023 +0900

    .gitlab-ci.yml: disable p11-kit for MinGW builds
    
    Currently, tests are failing with page fault in MinGW64 builds if
    p11-kit is enabled.  This temporarily disables it.
    
    Signed-off-by: Daiki Ueno <ueno@gnu.org>

Author: Daiki Ueno <ueno@gnu.org>
Date:   Fri Oct 13 17:36:57 2023 +0900

    cli: add --starttls-name option
    
    Some deployment of application protocols, such as XMPP, require a
    different hostname than the host being connected.  This adds a new
    option, --starttls-name, to gnutls-cli to specify it separately.
    
    Signed-off-by: Daiki Ueno <ueno@gnu.org>

Author: Daiki Ueno <ueno@gnu.org>
Date:   Sat Oct 7 17:25:45 2023 +0900

    .gitlab-ci.yml: hook devel/check-headers.sh in .pre target
    
    This adds devel/check-headers.sh in the .pre target phase so any
    inconsistent header inclusion would be detected early in the CI.
    
    Signed-off-by: Daiki Ueno <ueno@gnu.org>

Author: Daiki Ueno <ueno@gnu.org>
Date:   Sat Oct 7 17:24:02 2023 +0900

    devel: make local header inclusion consistent
    
    This is the result of running devel/check-headers.sh --format on all
    .c and .h files under lib.
    
    Signed-off-by: Daiki Ueno <ueno@gnu.org>

Author: Daiki Ueno <ueno@gnu.org>
Date:   Sat Oct 7 17:17:39 2023 +0900

    devel: add script to fix header inclusion
    
    This adds check-headers.sh script which scans .c files and checks that
    local header files under lib/ are consistently included with #include
    "foo.h" instead of #include <foo.h>.
    
    Signed-off-by: Daiki Ueno <ueno@gnu.org>

Author: Yongye Zhu <zyy1102000@gmail.com>
Date:   Wed Oct 4 09:26:14 2023 +0000

    ktls: add support for FreeBSD
    
    This extend the KTLS support to FreeBSD, with the AES-GCM-128, AES-GCM-256, and ChaCha20-Poly1305 ciphersuites.
    
    Signed-off-by: Yongye Zhu <zyy1102000@gmail.com>
    Reviewed-by: Frantisek Krenzelok <krenzelok.frantisek@gmail.com>

Author: Daiki Ueno <ueno@gnu.org>
Date:   Thu Sep 28 11:38:36 2023 +0900

    pkcs11: support Ed448 keys
    
    This adds support for Ed448 keys backed by PKCS#11.  To differentiate
    Ed448 keys from Ed25519 keys, this requires an extra logic to check
    CKA_EC_PARAMS when reading public keys.
    
    Signed-off-by: Daiki Ueno <ueno@gnu.org>

Author: Daiki Ueno <ueno@gnu.org>
Date:   Thu Sep 28 11:37:15 2023 +0900

    tests: refactor pkcs11-privkey-generate
    
    This factors out key generation logic into a separate function so we
    can easily extend the test to cover more key types.
    
    Signed-off-by: Daiki Ueno <ueno@gnu.org>

Author: Daiki Ueno <ueno@gnu.org>
Date:   Sat Sep 23 15:37:03 2023 +0900

    pkcs11: fix key pair generation for EdDSA
    
    Previouly we used the CKM_EDDSA mechanism to generate key pair, though
    the mechanism can only be used for signing and verification as
    specified in PKCS#11 3.1 section 6.3.  For key generation, the
    CKM_EC_EDWARDS_KEY_PAIR_GEN mechanism (or
    CKM_EC_MONTGOMERY_KEY_PAIR_GEN, if the point is represented in the
    Montgomery form) needs to be used.
    
    Signed-off-by: Daiki Ueno <ueno@gnu.org>

Author: Dimitri Papadopoulos <3350651-DimitriPapadopoulos@users.noreply.gitlab.com>
Date:   Sun Sep 17 16:29:38 2023 +0200

    doc: fix typos found by codespell
    
    Signed-off-by: Dimitri Papadopoulos <3350651-DimitriPapadopoulos@users.noreply.gitlab.com>

Author: Daiki Ueno <ueno@gnu.org>
Date:   Mon Sep 11 13:59:45 2023 +0900


src/srptool.c
/*
 * Copyright (C) 2001-2012 Free Software Foundation, Inc.
 *
 * This file is part of GnuTLS.
 *
 * GnuTLS is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * GnuTLS is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see
 * <https://www.gnu.org/licenses/>.
 */

#include "config.h"

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <gnutls/gnutls.h>
#include <gnutls/crypto.h> /* for random */

#include <sys/types.h>
#include <sys/stat.h>

#ifndef _WIN32
#include <pwd.h>
#include <unistd.h>
#else
#include <windows.h>
#endif

/* Gnulib portability files. */
#include <getpass.h>
#include <minmax.h>

#include "srptool-options.h"

/* This may need some rewrite. A lot of stuff which should be here
 * are in the library, which is not good.
 */

int crypt_int(const char *username, const char *passwd, int salt,
	      const char *tpasswd_conf, const char *tpasswd, int uindex);
static int read_conf_values(gnutls_datum_t *g, gnutls_datum_t *n, char *str);
static int _verify_passwd_int(const char *username, const char *passwd,
			      char *verifier, const char *salt,
			      const gnutls_datum_t *g, const gnutls_datum_t *n);

static void print_num(const char *msg, const gnutls_datum_t *num)
{
	unsigned int i;

	printf("%s:\t", msg);

	for (i = 0; i < num->size; i++) {
		if (i != 0 && i % 12 == 0)
			printf("\n\t");
		else if (i != 0 && i != num->size)
			printf(":");
		printf("%.2x", num->data[i]);
	}
	printf("\n\n");
}

static int generate_create_conf(const char *tpasswd_conf)
{
	FILE *fp;
	char line[5 * 1024];
	int index = 1, srp_idx;
	gnutls_datum_t g, n;
	gnutls_datum_t str_g, str_n;

	fp = fopen(tpasswd_conf, "w");
	if (fp == NULL) {
		fprintf(stderr, "Cannot open file '%s'\n", tpasswd_conf);
		return -1;
	}

	for (index = 1; index <= 5; index++) {
		if (index == 1) {
			srp_idx = 2;
			n = gnutls_srp_1536_group_prime;
			g = gnutls_srp_1536_group_generator;
		} else if (index == 2) {
			srp_idx = 3;
			n = gnutls_srp_2048_group_prime;
			g = gnutls_srp_2048_group_generator;
		} else if (index == 3) {
			srp_idx = 4;
			n = gnutls_srp_3072_group_prime;
			g = gnutls_srp_3072_group_generator;
		} else if (index == 4) {
			srp_idx = 5;
			n = gnutls_srp_4096_group_prime;
			g = gnutls_srp_4096_group_generator;
		} else if (index == 5) {
			srp_idx = 7;
			n = gnutls_srp_8192_group_prime;
			g = gnutls_srp_8192_group_generator;
		} else {
			fprintf(stderr, "Unknown index: %d\n", index);
			fclose(fp);
			return -1;
		}

		printf("\nGroup %d, of %d bits:\n", srp_idx, n.size * 8);
		print_num("Generator", &g);
		print_num("Prime", &n);

		if (gnutls_srp_base64_encode_alloc(&n, &str_n) < 0) {
			fprintf(stderr, "Could not encode\n");
			fclose(fp);
			return -1;
		}

		if (gnutls_srp_base64_encode_alloc(&g, &str_g) < 0) {
			fprintf(stderr, "Could not encode\n");
			fclose(fp);
			return -1;
		}

		sprintf(line, "%d:%s:%s\n", srp_idx, str_n.data, str_g.data);

		gnutls_free(str_n.data);
		gnutls_free(str_g.data);

		fwrite(line, 1, strlen(line), fp);
	}

	fclose(fp);

	return 0;
}

/* The format of a tpasswd file is:
 * username:verifier:salt:index
 *
 * index is the index of the prime-generator pair in tpasswd.conf
 */
static int _verify_passwd_int(const char *username, const char *passwd,
			      char *verifier, const char *salt,
			      const gnutls_datum_t *g, const gnutls_datum_t *n)
{
	char _salt[1024];
	gnutls_datum_t tmp, raw_salt, new_verifier;
	size_t salt_size;
	char *pos;

	if (salt == NULL || verifier == NULL)
		return -1;

	if (strlen(salt) >= sizeof(_salt)) {
		fprintf(stderr, "Too long salt.\n");
		return -1;
	}

	/* copy salt, and null terminate after the ':' */
	strcpy(_salt, salt);
	pos = strchr(_salt, ':');
	if (pos != NULL)
		*pos = 0;

	/* convert salt to binary. */
	tmp.data = (void *)_salt;
	tmp.size = strlen(_salt);

	if (gnutls_srp_base64_decode_alloc(&tmp, &raw_salt) < 0) {
		fprintf(stderr, "Could not decode salt.\n");
		return -1;
	}

	if (gnutls_srp_verifier(username, passwd, &raw_salt, g, n,
				&new_verifier) < 0) {
		fprintf(stderr, "Could not make the verifier\n");
		return -1;
	}

	free(raw_salt.data);

	/* encode the verifier into _salt */
	salt_size = sizeof(_salt);
	memset(_salt, 0, salt_size);
	if (gnutls_srp_base64_encode(&new_verifier, _salt, &salt_size) < 0) {
		fprintf(stderr, "Encoding error\n");
		return -1;
	}

	free(new_verifier.data);

	if (strncmp(verifier, _salt, strlen(_salt)) == 0) {
		fprintf(stderr, "Password verified\n");
		return 0;
	} else {
		fprintf(stderr, "Password does NOT match\n");
	}
	return -1;
}

static int filecopy(const char *src, const char *dst)
{
	FILE *fp, *fp2;
	char line[5 * 1024];
	char *p;

	fp = fopen(dst, "w");
	if (fp == NULL) {
		fprintf(stderr, "Cannot open '%s' for write\n", dst);
		return -1;
	}

	fp2 = fopen(src, "r");
	if (fp2 == NULL) {
		/* empty file */
		fclose(fp);
		return 0;
	}

	line[sizeof(line) - 1] = 0;
	do {
		p = fgets(line, sizeof(line) - 1, fp2);
		if (p == NULL)
			break;

		fputs(line, fp);
	} while (1);

	fclose(fp);
	fclose(fp2);

	return 0;
}

/* accepts password file */
static int find_strchr(const char *username, const char *file)
{
	FILE *fp;
	char line[5 * 1024];
	size_t len = strlen(username);

	fp = fopen(file, "r");
	if (fp == NULL) {
		fprintf(stderr, "Cannot open file '%s'\n", file);
		return -1;
	}

	while (fgets(line, sizeof(line), fp) != NULL) {
		/* parse entry in the form: <username>:<crypt>:<index> */
		if (strncmp(username, line, len) == 0 && line[len] == ':') {
			char *pos;

			/* find the index */
			pos = strrchr(line, ':');
			pos++;
			fclose(fp);
			return atoi(pos);
		}
	}

	fclose(fp);
	return -1;
}

/* Parses the tpasswd files, in order to verify the given
 * username/password pair.
 */
static int verify_passwd(const char *conffile, const char *tpasswd,
			 const char *username, const char *passwd)
{
	FILE *fp;
	char line[5 * 1024];
	gnutls_datum_t g, n;
	int iindex;
	char *p;
	size_t len = strlen(username);

	iindex = find_strchr(username, tpasswd);
	if (iindex == -1) {
		fprintf(stderr, "Cannot find '%s' in %s\n", username, tpasswd);
		return -1;
	}

	fp = fopen(conffile, "r");
	if (fp == NULL) {
		fprintf(stderr, "Cannot find %s\n", conffile);
		return -1;
	}

	do {
		p = fgets(line, sizeof(line) - 1, fp);
	} while (p != NULL && atoi(p) != iindex);

	fclose(fp);

	if (p == NULL) {
		fprintf(stderr, "Cannot find entry in %s\n", conffile);
		return -1;
	}
	line[sizeof(line) - 1] = 0;

	if (read_conf_values(&g, &n, line) < 0) {
		fprintf(stderr, "Cannot parse conf file '%s'\n", conffile);
		return -1;
	}

	fp = fopen(tpasswd, "r");
	if (fp == NULL) {
		fprintf(stderr, "Cannot open file '%s'\n", tpasswd);
		return -1;
	}

	while (fgets(line, sizeof(line), fp) != NULL) {
		/* parse entry in the form: <username>:<crypt>:<index> */
		if (strncmp(username, line, len) == 0 && line[len] == ':') {
			char *verifier_pos, *salt_pos, *pos;

			fclose(fp);

			pos = &line[len + 1];
			verifier_pos = pos;

			/* Move to the salt */
			pos = strchr(pos, ':');
			if (pos == NULL) {
				fprintf(stderr, "Cannot parse conf file '%s'\n",
					conffile);
				return -1;
			}
			pos++;
			salt_pos = pos;

			return _verify_passwd_int(username, passwd,
						  verifier_pos, salt_pos, &g,
						  &n);
		}
	}

	fclose(fp);
	return -1;
}

#ifdef __OS2__
#define KPASSWD "/@unixroot/etc/tpasswd"
#define KPASSWD_CONF "/@unixroot/etc/tpasswd.conf"
#else
#define KPASSWD "/etc/tpasswd"
#define KPASSWD_CONF "/etc/tpasswd.conf"
#endif

static void tls_log_func(int level, const char *str)
{
	fprintf(stderr, "|<%d>| %s", level, str);
}

int main(int argc, char **argv)
{
	const char *passwd;
	int salt_size, ret;
	const char *fpasswd, *fpasswd_conf;
	const char *username;
#ifndef _WIN32
	struct passwd *pwd;
#endif

	if ((ret = gnutls_global_init()) < 0) {
		fprintf(stderr, "global_init: %s\n", gnutls_strerror(ret));
		exit(1);
	}

	umask(066);

	optionProcess(&srptoolOptions, argc, argv);

	gnutls_global_set_log_function(tls_log_func);
	gnutls_global_set_log_level(OPT_VALUE_DEBUG);

	if (HAVE_OPT(CREATE_CONF)) {
		return generate_create_conf(OPT_ARG(CREATE_CONF));
	}

	if (HAVE_OPT(PASSWD))
		fpasswd = OPT_ARG(PASSWD);
	else
		fpasswd = (char *)KPASSWD;

	if (HAVE_OPT(PASSWD_CONF))
		fpasswd_conf = OPT_ARG(PASSWD_CONF);
	else
		fpasswd_conf = (char *)KPASSWD_CONF;

	if (HAVE_OPT(USERNAME))
		username = OPT_ARG(USERNAME);
	else {
#ifndef _WIN32
		pwd = getpwuid(getuid());

		if (pwd == NULL) {
			fprintf(stderr, "No such user\n");
			return -1;
		}

		username = pwd->pw_name;
#else
		fprintf(stderr, "Please specify a user\n");
		return -1;
#endif
	}

	salt_size = 16;

	passwd = getpass("Enter password: ");
	if (passwd == NULL) {
		fprintf(stderr, "Please specify a password\n");
		return -1;
	}

	/* not ready yet */
	if (HAVE_OPT(VERIFY)) {
		return verify_passwd(fpasswd_conf, fpasswd, username, passwd);
	}

	return crypt_int(username, passwd, salt_size, fpasswd_conf, fpasswd,
			 OPT_VALUE_INDEX);
}

static char *_srp_crypt(const char *username, const char *passwd, int salt_size,
			const gnutls_datum_t *g, const gnutls_datum_t *n)
{
	unsigned char salt[128];
	static char result[1024];
	gnutls_datum_t dat_salt, txt_salt;
	gnutls_datum_t verifier, txt_verifier;

	if ((unsigned)salt_size > sizeof(salt))
		return NULL;

	/* generate the salt
	 */
	if (gnutls_rnd(GNUTLS_RND_NONCE, salt, salt_size) < 0) {
		fprintf(stderr, "Could not create nonce\n");
		return NULL;
	}

	dat_salt.data = salt;
	dat_salt.size = salt_size;

	if (gnutls_srp_verifier(username, passwd, &dat_salt, g, n, &verifier) <
	    0) {
		fprintf(stderr, "Error getting verifier\n");
		return NULL;
	}

	/* base64 encode the verifier */
	if (gnutls_srp_base64_encode_alloc(&verifier, &txt_verifier) < 0) {
		fprintf(stderr, "Error encoding\n");
		free(verifier.data);
		return NULL;
	}

	free(verifier.data);

	if (gnutls_srp_base64_encode_alloc(&dat_salt, &txt_salt) < 0) {
		fprintf(stderr, "Error encoding\n");
		return NULL;
	}

	sprintf(result, "%s:%s", txt_verifier.data, txt_salt.data);
	free(txt_salt.data);
	free(txt_verifier.data);

	return result;
}

int crypt_int(const char *username, const char *passwd, int salt_size,
	      const char *tpasswd_conf, const char *tpasswd, int uindex)
{
	FILE *fp;
	char *cr;
	gnutls_datum_t g, n;
	char line[5 * 1024];
	char *p, *pp;
	int iindex;
	char tmpname[1024];

	fp = fopen(tpasswd_conf, "r");
	if (fp == NULL) {
		fprintf(stderr, "Cannot find %s\n", tpasswd_conf);
		return -1;
	}

	do { /* find the specified uindex in file */
		p = fgets(line, sizeof(line) - 1, fp);
	} while (p != NULL && atoi(p) != uindex);
	fclose(fp);

	if (p == NULL) {
		fprintf(stderr, "Cannot find entry in %s\n", tpasswd_conf);
		return -1;
	}
	line[sizeof(line) - 1] = 0;

	if ((iindex = read_conf_values(&g, &n, line)) < 0) {
		fprintf(stderr, "Cannot parse conf file '%s'\n", tpasswd_conf);
		return -1;
	}

	cr = _srp_crypt(username, passwd, salt_size, &g, &n);
	if (cr == NULL) {
		fprintf(stderr, "Cannot _srp_crypt()...\n");
		return -1;
	} else {
		/* delete previous entry */
		struct stat st;
		FILE *fp2;
		int put;

		if (strlen(tpasswd) + 5 > sizeof(tmpname)) {
			fprintf(stderr, "file '%s' is tooooo long\n", tpasswd);
			return -1;
		}

		snprintf(tmpname, sizeof(tmpname), "%s.tmp", tpasswd);

		if (stat(tmpname, &st) != -1) {
			fprintf(stderr, "file '%s' is locked\n", tpasswd);
			return -1;
		}

		if (filecopy(tpasswd, tmpname) != 0) {
			fprintf(stderr, "Cannot copy '%s' to '%s'\n", tpasswd,
				tmpname);
			return -1;
		}

		fp = fopen(tpasswd, "w");
		if (fp == NULL) {
			fprintf(stderr, "Cannot open '%s' for write\n",
				tpasswd);
			(void)remove(tmpname);
			return -1;
		}

		fp2 = fopen(tmpname, "r");
		if (fp2 == NULL) {
			fprintf(stderr, "Cannot open '%s' for read\n", tmpname);
			(void)remove(tmpname);
			fclose(fp);
			return -1;
		}

		put = 0;
		do {
			p = fgets(line, sizeof(line) - 1, fp2);
			if (p == NULL)
				break;

			pp = strchr(line, ':');
			if (pp == NULL)
				continue;

			if (strncmp(p, username,
				    MAX(strlen(username),
					(unsigned int)(pp - p))) == 0) {
				put = 1;
				fprintf(fp, "%s:%s:%u\n", username, cr, iindex);
			} else {
				fputs(line, fp);
			}
		} while (1);

		if (put == 0) {
			fprintf(fp, "%s:%s:%u\n", username, cr, iindex);
		}

		fclose(fp);
		fclose(fp2);

		(void)remove(tmpname);
	}

	return 0;
}

/* this function parses tpasswd.conf file. Format is:
 * int(index):base64(n):base64(g)
 */
static int read_conf_values(gnutls_datum_t *g, gnutls_datum_t *n, char *str)
{
	char *p;
	int len;
	int index, ret;
	gnutls_datum_t dat;

	index = atoi(str);

	p = strrchr(str, ':'); /* we have g */
	if (p == NULL) {
		return -1;
	}

	*p = '\0';
	p++;

	/* read the generator */
	len = strlen(p);
	if (p[len - 1] == '\n')
		len--;

	dat.data = (void *)p;
	dat.size = len;
	ret = gnutls_srp_base64_decode_alloc(&dat, g);

	if (ret < 0) {
		fprintf(stderr, "Decoding error\n");
		return -1;
	}

	/* now go for n - modulo */
	p = strrchr(str, ':'); /* we have n */
	if (p == NULL) {
		return -1;
	}

	*p = '\0';
	p++;

	dat.data = (void *)p;
	dat.size = strlen(p);

	ret = gnutls_srp_base64_decode_alloc(&dat, n);

	if (ret < 0) {
		fprintf(stderr, "Decoding error\n");
		free(g->data);
		return -1;
	}

	return index;
}
